src/renderer/dx/DxSoftFont.cpp

```diff
@@ -0,0 +1,256 @@
+// the dimensions for a typical 2:1 cell size. Each position in the grid is
+// surrounded by a 2 pixel border which helps avoid bleed across the character
+// boundaries when the output is scaled.
+
+constexpr size_t BITMAP_GRID_WIDTH = 12;
+constexpr size_t BITMAP_GRID_HEIGHT = 8;
+constexpr size_t PADDING = 2;
+
+constexpr auto ANTIALIASED_INTERPOLATION = D2D1_SCALE_INTERPOLATION_MODE_HIGH_QUALITY_CUBIC;
+constexpr auto ALIASED_INTERPOLATION = D2D1_SCALE_INTERPOLATION_MODE_NEAREST_NEIGHBOR;
+
+DxSoftFont::DxSoftFont() noexcept :
+    _centeringHint{},
+    _interpolation{ ANTIALIASED_INTERPOLATION },
+    _colorMatrix{}
+{
+    _colorMatrix.m[0][3] = 1;
+}
+
+void DxSoftFont::SetFont(const gsl::span<const uint16_t> bitPattern,
+                         const til::size sourceSize,
+                         const til::size targetSize,
+                         const size_t centeringHint)
+{
+    Reset();
+
+    // If the font is being reset, just free up the memory and return.
+    if (bitPattern.empty())
+    {
+        _bitmapBits.clear();
+        return;
+    }
+
+    const auto maxGlyphCount = BITMAP_GRID_WIDTH * BITMAP_GRID_HEIGHT;
+    _glyphCount = std::min<size_t>(bitPattern.size() / sourceSize.height, maxGlyphCount);
+    _sourceSize = sourceSize;
+    _targetSize = targetSize;
+    _centeringHint = centeringHint;
+
+    const auto bitmapWidth = BITMAP_GRID_WIDTH * (_sourceSize.width + PADDING * 2);
+    const auto bitmapHeight = BITMAP_GRID_HEIGHT * (_sourceSize.height + PADDING * 2);
+    _bitmapBits = std::vector<UINT32>(bitmapWidth * bitmapHeight);
+    _bitmapSize = { gsl::narrow_cast<UINT32>(bitmapWidth), gsl::narrow_cast<UINT32>(bitmapHeight) };
+
+    const auto bitmapScanline = [=](const auto lineNumber) noexcept {
+        return _bitmapBits.begin() + lineNumber * bitmapWidth;
+    };
+
+    // The source bitPattern is just a list of the scanlines making up the
+    // glyphs one after the other, but we want to lay them out in a grid, so
+    // we need to process each glyph individually.
+    auto srcPointer = bitPattern.begin();
+    for (auto glyphNumber = 0u; glyphNumber < _glyphCount; glyphNumber++)
+    {
+        // We start by calculating the position in the bitmap where the glyph
+        // needs to be stored.
+        const auto xOffset = _xOffsetForGlyph<size_t>(glyphNumber);
+        const auto yOffset = _yOffsetForGlyph<size_t>(glyphNumber);
+        auto dstPointer = bitmapScanline(yOffset) + xOffset;
+        for (auto y = 0; y < sourceSize.height; y++)
+        {
+            // Then for each scanline in the source, we need expand the bits
+            // into 32-bit BGRA values. For every bit that is set we write out
+            // an FFFFFFFF value, and it not set, we write out 00000000. In the
+            // end, though, all we care about is the red component, since we'll
+            // later remap these values with a color matrix.

```

Comment:
If you need a monochrome bitmap why not use `DXGI_FORMAT_R8_UNORM` or `DXGI_FORMAT_A8_UNORM`?
The latter would work like this: https://docs.microsoft.com/en-us/windows/win32/direct2d/how-to-clip-with-bitmap-masks

Reply:
OK, I've just tried with `DXGI_FORMAT_R8_UNORM` now and that appears to work fine. I have no idea what I was doing wrong before. I think I might have originally been using it as a render target, so maybe that was the difference. Anyway I'll update it now. Thanks for bringing that up.